### PR TITLE
Refactor ValuesNode

### PR DESCRIPTION
**Pull request by kasiafi**
Change inner rows structure of `ValuesNode` from `List<List<Expression>>`
to `List<Expression>`.
Fix the issue of Values rows specified by expressions other than Row
but evaluating to RowType.
In that case, Values node failed at Planning, as the expression
was considered a single field specification instead of
row specification.

Fixes https://github.com/prestosql/presto/issues/3398

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneValuesColumns.java [hunk 1]
```diff
@@ -39,6 +41,17 @@ public PruneValuesColumns()
...
     protected Optional<PlanNode> pushDownProjectOff(Context context, ValuesNode valuesNode, Set<Symbol> referencedOutputs)
     {
+        // no symbols to prune
+        if (valuesNode.getOutputSymbols().isEmpty()) {
+            return Optional.empty();
+        }
+
+        checkState(valuesNode.getRows().isPresent(), "rows is empty");
+        // if any of ValuesNode's rows is specified by expression other than Row, the redundant piece cannot be extracted and pruned
+        if (valuesNode.getRows().get().stream().anyMatch(row -> !(row instanceof Row))) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneValuesColumns.java:51]: Alternatively:

```java
        if (!valuesNode.getRows().get().stream().allMatch(Row.class::isInstance)) {
```

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java [hunk 2]
```diff
@@ -763,10 +764,20 @@ private PlanNode rewriteSetOperationChildren(SetOperationNode node, RewriteConte
         @Override
         public PlanNode visitValues(ValuesNode node, RewriteContext<Set<Symbol>> context)
         {
+            // nothing to prune: no output symbols and no expressions
+            if (node.getRows().isEmpty()) {
+                return node;
+            }
+
+            // if any of ValuesNode's rows is specified by expression other than Row, the redundant piece cannot be extracted and pruned
+            if (node.getRows().get().stream().anyMatch(row -> !(row instanceof Row))) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java:773]: Alternatively:

```java
        if (!node.getRows().get().stream().allMatch(Row.class::isInstance)) {
```

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 3]
```diff
@@ -462,11 +463,34 @@ public PlanAndMappings visitValues(ValuesNode node, UnaliasContext context)
...
+                List<Symbol> newOutputs = node.getOutputSymbols().stream()
+                        .map(mapper::map)
+                        .distinct()
+                        .collect(toImmutableList());
+                checkState(newOutputs.size() == node.getOutputSymbols().size(), "duplicate output symbol in Values");
+                return new PlanAndMappings(
+                        new ValuesNode(
+                                node.getId(),
+                                newOutputs,
+                                newRows),
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/UnaliasSymbolReferences.java:485]: Format as:

```java
new ValuesNode(node.getId(), newOutputs, newRows)
```

File: presto-main/src/main/java/io/prestosql/sql/planner/planprinter/PlanPrinter.java [hunk 4]
```diff
@@ -697,11 +698,25 @@ public Void visitTableScan(TableScanNode node, Void context)
...
+                return null;
+            }
+            List<String> rows = node.getRows().get().stream()
+                    .map(row -> {
+                        if (row instanceof Row) {
+                            return ((Row) row).getItems().stream()
+                                    .map(PlanPrinter::unresolveFunctions)
+                                    .map(Expression::toString)
+                                    .collect(joining(", ", "(", ")"));
+                        }
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/planprinter/PlanPrinter.java:714]: Why is this needed? The expression printer should be able to print a `Row` expression directly.

Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/planprinter/PlanPrinter.java:714]: I wanted to preserve the result for all "regular" rows.
`Row` expression renders as "ROW(..."

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/ValuesNode.java [hunk 5]
```diff
@@ -18,33 +18,85 @@
...
         extends PlanNode
 {
     private final List<Symbol> outputSymbols;
-    private final List<List<Expression>> rows;
+    private final int rowCount;
+    // If ValuesNode produces output symbols, each row in ValuesNode is represented by a single expression in `rows` list.
+    // It can be an expression of type Row or any other expression that evaluates to RowType.
+    // In case when output symbols are present but ValuesNode does not have any rows, `rows` is an Optional with empty list.
+    // If ValuesNode does not produce any output symbols, `rows` is Optional.empty().
+    private final Optional<List<Expression>> rows;
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/plan/ValuesNode.java:44]: As long as it is `List<Expression>`, I can see no better way than to wrap it in Optional. 
At first, I tried to create `Row` with no fields. Even though it is legal to create, it fails to evaluate.

[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/plan/ValuesNode.java:44]: Yes, I know. Not asking for any changes -- just commenting on what would be desirable in the longer term.

File: presto-main/src/main/java/io/prestosql/sql/planner/EffectivePredicateExtractor.java [hunk 6]
```diff
@@ -336,69 +340,103 @@ public Expression visitValues(ValuesNode node, Void context)
...
-                Type type = types.get(symbol);
+            // for each row of Values, get all expressions that will be evaluated:
+            // - if the row is of type Row, evaluate fields of the row
+            // - otherwise evaluate the whole expression and then analyze fields of the resulting row
+            checkState(node.getRows().isPresent(), "rows is empty");
+            List<Expression> processedExpressions = node.getRows().get().stream()
+                    .flatMap(row -> {
+                        if (row instanceof Row) {
+                            return ((Row) row).getItems().stream();
+                        }
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/EffectivePredicateExtractor.java:351]: I wanted to take advantage of analyzing Row items separately in case some of them are non-deterministic and other are deterministic. When analyzing them one by one, we still can build domains for the deterministic columns.

File: presto-main/src/main/java/io/prestosql/sql/planner/EffectivePredicateExtractor.java [hunk 7]
```diff
@@ -336,69 +340,103 @@ public Expression visitValues(ValuesNode node, Void context)
...
                     }
-
-                    ExpressionInterpreter interpreter = ExpressionInterpreter.expressionOptimizer(value, metadata, session, expressionTypes);
+                }
+                else {
+                    ExpressionInterpreter interpreter = ExpressionInterpreter.expressionOptimizer(row, metadata, session, expressionTypes);
                     Object evaluated = interpreter.optimize(NoOpSymbolResolver.INSTANCE);
-
                     if (evaluated instanceof Expression) {
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/EffectivePredicateExtractor.java:399]: The same treatment as above means that we return TRUE_LITERAL from the method if we find any unresolved item. I just copied this behavior for non-Row expressions. 
The predicate extractor could be further improved to only skip a single column in such case.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java [hunk 8]
```diff
@@ -285,18 +285,27 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
...
-                    newRow.add(rewritten);
+            ImmutableList.Builder<Expression> rows = ImmutableList.builder();
+            for (Expression row : valuesNode.getRows().get()) {
+                Expression rewritten;
+                if (row instanceof Row) {
+                    // preserve the structure of row
+                    rewritten = new Row(((Row) row).getItems().stream()
+                            .map(item -> rewriter.rewrite(item, context))
+                            .collect(toImmutableList()));
+                }
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java:301]: Yes... I found I needed to extract the case of Row expression because of `SimplifyExpressions`.
It uses the `LiteralEncoder` which turns a Row into something else. I had to process the Row by field and then put it back together so that it remained a `Row` at the top level.

**Comment by kasiafi** [comment]
Applied comments. Also, improved column pruning for ValuesNode.
